include/geometric_shapes/random_number_utils.hpp

```diff
@@ -0,0 +1,112 @@
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+/* Author: Sebastian Jahr
+   Description: A random number generator singleton implementation
+*/
+
+#pragma once
+
+#include <random>       // for std::mt19937, random_device , seed_seq
+#include <array>        // for std::array, std::data,
+#include <algorithm>    // for std::generate_n
+#include <functional>   // for std::ref
+#include "Eigen/Dense"  // for Eigen::Quaterniond::UnitRandom();
+#include <time.h>       // for time
+#include <optional>     // for optional
+
+namespace shapes
+{
+// Singleton random number generator class based on https://www.modernescpp.com/index.php/thread-safe-initialization-of-a-singleton
+class RandomNumberGenerator
+{
+private:
+  std::mt19937 generator_;
+
+  // Delete copy constructor
+  RandomNumberGenerator(const RandomNumberGenerator&) = delete;
+  // Don't allow assigning the instance of this class
+  RandomNumberGenerator& operator=(const RandomNumberGenerator&) = delete;
+
+  RandomNumberGenerator()
+    : generator_{ []() {
+      std::array<int, std::mt19937::state_size> seed_data;
+      std::random_device random_device;
+      std::generate_n(std::data(seed_data), std::size(seed_data), std::ref(random_device));
+      std::seed_seq sequence(std::begin(seed_data), std::end(seed_data));
+      std::mt19937 generator(sequence);
+      return generator;
+    }() }
+  {
+    // Seed srand initially
+    std::srand(time(NULL));
+  };
+
+  RandomNumberGenerator(std::seed_seq& seed_sequence)
+    : generator_{ [](std::seed_seq& seed_sequence) {
+      std::mt19937 generator(seed_sequence);
+      return generator;
+    }(seed_sequence) } {};
+  ~RandomNumberGenerator() = default;
+
+public:
+  [[nodiscard]] static RandomNumberGenerator& getInstance(std::optional<std::seed_seq> seed_sequence = std::nullopt)
+  {
+    // Static valiables with blocked scopes will be only created once
+    if (seed_sequence.has_value())
+    {
+      thread_local RandomNumberGenerator instance(seed_sequence.value());
+      return instance;
+    }
+    else
+    {
+      thread_local RandomNumberGenerator instance;
+      return instance;
+    }
+  }
+
+  /// \brief Return a random number based on a bounded uniform distribution
+  [[nodiscard]] auto uniform(double const lower_bound, double const upper_bound)
+  {
+    std::uniform_real_distribution<double> distribution(lower_bound, upper_bound);
+    return distribution(generator_);
+  }
+
+  /// \brief Return a random quaternion. The reason for this wrapper function is that we need to make sure, that srand()
+  /// is seeded before UnitRandom() is called. The seeding takes place in the Singleton's constructor.
+  [[nodiscard]] auto getRandomQuaternion(std::optional<unsigned int> seed = std::nullopt)
+  {
+    if (seed.has_value())
+    {
+      // Optional re-seed
+      std::srand(seed.value());
+    }
```

Comment:
Why would you want to re-seed here?

Reply:
Initially, srand is seeded with NULL but for testing or other cases where a deterministic random number is required I've added this option